connector/controllers/main.py

```diff
@@ -60,28 +86,17 @@ def retry_postpone(job, message, seconds=None):
                 return ""
 
         try:
-            # if the job has been manually set to DONE or PENDING,
-            # or if something tries to run a job that is not enqueued
-            # before its execution, stop
-            if job.state != ENQUEUED:
-                _logger.warning('job %s is in state %s '
-                                'instead of enqueued in /runjob',
-                                job_uuid, job.state)
-                return
-
-            with session_hdl.session() as session:
-                # TODO: set_started should be done atomically with
-                #       update queue_job set=state=started
-                #       where state=enqueid and id=
-                job.set_started()
-                self.job_storage_class(session).store(job)
-
-            _logger.debug('%s started', job)
-            with session_hdl.session() as session:
-                job.perform(session)
-                job.set_done()
-                self.job_storage_class(session).store(job)
-            _logger.debug('%s done', job)
+            try:
+                self._try_perform_job(session_hdl, job)
+            except OperationalError as err:
+                # Automatically retry the typical transaction serialization
+                # errors
+                if err.pgcode not in PG_CONCURRENCY_ERRORS_TO_RETRY:
+                    raise
+
+                session.env.clear()
```

Comment:
Thinking again, I think your proposition to use `clear_upon_failure` in `ConnectorSessionHandler` is better and should cover all places where this error can occur.

```python
    @contextmanager
    def session(self):
        """ Context Manager: start a new session and ensure that the
        session's cursor is:

        * rollbacked on errors
        * commited at the end of the ``with`` context when no error occured
        * always closed at the end of the ``with`` context
        * it handles the registry signaling
        """
        with openerp.api.Environment.manage():
            db = openerp.sql_db.db_connect(self.db_name)
            session = ConnectorSession(db.cursor(),
                                       self.uid,
                                       context=self.context)

            try:
                with session.env.clear_upon_failure():
                    RegistryManager.check_registry_signaling(self.db_name)
                    yield session
                    RegistryManager.signal_caches_change(self.db_name)
            except:
                session.rollback()
                raise
            else:
                session.commit()
            finally:
                session.close()


```